src/macaron/slsa_analyzer/provenance/intoto/v1/__init__.py

```diff
@@ -12,3 +44,211 @@ class InTotoV1Statement(TypedDict):
+            "The value of attribute '_type' in the in-toto statement must be: 'https://in-toto.io/Statement/v1'",
+        )
+
+    subjects_payload = payload.get("subject")
+    if subjects_payload is None:
+        raise ValidateInTotoPayloadError(
+            "The attribute 'subject' of the in-toto statement is missing.",
+        )
+    if not isinstance(subjects_payload, list):
+        raise ValidateInTotoPayloadError(
+            "The value of attribute 'subject' in the in-toto statement is invalid: expecting a list.",
+        )
+
+    for subject_json in subjects_payload:
+        validate_intoto_subject(subject_json)
+
+    predicate_type = payload.get("predicateType")
+    if predicate_type is None:
+        raise ValidateInTotoPayloadError(
+            "The attribute 'predicateType' of the in-toto statement is missing.",
+        )
+
+    if not isinstance(predicate_type, str):
+        raise ValidateInTotoPayloadError(
+            "The value of attribute 'predicateType' in the in-toto statement is invalid: expecting a string."
+        )
+
+    predicate = payload.get("predicate")
+    if predicate is not None and not isinstance(predicate, dict):
+        raise ValidateInTotoPayloadError(
+            "The value attribute 'predicate' in the in-toto statement is invalid: expecting an object.",
+        )
+
+    return True
+
+
+def validate_intoto_subject(subject: JsonType) -> TypeGuard[InTotoV1ResourceDescriptor]:
+    """Validate a single subject in the in-toto statement.
+
+    See specification: https://github.com/in-toto/attestation/blob/main/spec/v1/resource_descriptor.md
+
+    Parameters
+    ----------
+    subject : JsonType
+        The JSON element representing a single subject.
+
+    Returns
+    -------
+    TypeGuard[InTotoV1ResourceDescriptor]
+        ``True`` if the subject element is valid, in which case its type is narrowed to an
+        ``InTotoV1ResourceDescriptor``; ``False`` otherwise.
+
+    Raises
+    ------
+    ValidateInTotoPayloadError
+        When the payload does not follow the expecting schema.
+    """
+    if not isinstance(subject, dict):
+        raise ValidateInTotoPayloadError(
+            "A subject in the in-toto statement is invalid: expecting an object.",
+        )
+
+    # At least one of 'uri', 'digest', and 'content' must be valid and present.
+    uri_state = _validate_property(subject, "uri", True, lambda x: isinstance(x, str), False)
+    content_state = _validate_property(subject, "content", True, lambda x: isinstance(x, str), False)
+    digest_state = _validate_property(subject, "digest", True, is_valid_digest_set, False)
+    if _ValidityState.VALID not in (uri_state, content_state, digest_state):
+        raise ValidateInTotoPayloadError(
+            f"One of 'uri', 'digest', or 'content' must be present and valid within 'subject'. URI: {uri_state.name}. "
+            f"Content: {content_state.name}. DIGEST: {digest_state.name}."
+        )
```

Comment:
I guess something else you can do instead of using this `report_errors` param, is to return from `_validate_property` the object you are validating (if it is not there, return `None`). Then you can use the values to check for the co-existence of attributes.
```suggestion
    # At least one of 'uri', 'digest', and 'content' must be valid and present.
    uri = _validate_property(subject, "uri", required=False, validator=lambda x: isinstance(x, str))
    content = _validate_property(subject, "content", required=False, validator=lambda x: isinstance(x, str))
    digest = _validate_property(subject, "digest", required=False, validator=is_valid_digest_set)
    if not any([uri, content, digest]):
        raise ValidateInTotoPayloadError("One of 'uri', 'digest', or 'content' must be present and valid within 'subject'")
```

The `_validate_property` function then can be like this:

```python
def _validate_property(
    object_: dict[str, JsonType],
    key: str,
    required: bool,
    validator: Callable[[JsonType], bool],
) -> _ValidityState:
    """Validate the existence and type of target within the passed Json object."""
    value = object_.get(key)

    if not value:
        if required:
            raise ValidateInTotoPayloadError(f"The attribute {key} of the in-toto subject is missing.")
        return None

    if not validator(value):
        raise ValidateInTotoPayloadError(f"The attribute {key} of the in-toto subject is invalid.")

    return value
```